googlespanner/src/main/java/com/yahoo/ycsb/db/GoogleSpannerClient.java

```diff
@@ -0,0 +1,185 @@
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package com.yahoo.ycsb.db;
+
+import com.google.cloud.spanner.*;
+import com.yahoo.ycsb.ByteIterator;
+import com.yahoo.ycsb.DB;
+import com.yahoo.ycsb.DBException;
+import com.yahoo.ycsb.Status;
+import com.yahoo.ycsb.StringByteIterator;
+
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Google Cloud Spanner Client for YCSB.
+ *
+ * To use it, manually create an instance and a database with a table (e.g. in the cloud console). The
+ * table and database can have any name, but columns must be "pkey" (string, primary key), and
+ * "field0", "field1", .... "fieldN" depending on the number of fields in the workload.
+ *
+ * At minimum specify these properties:
+ *
+ * <ul>
+ *     <li> googlespanner.instance - name of your spanner instance</li>
+ *     <li> googlespanner.database - name of your spanner database</li>
+ *     <li> table - name of the table you created in spanner</li>
+ * </ul>
+ *
+ * This binding assumes application default credentials and cannot currently be used with any other authentication
+ * mechanism.
+ */
+
+public class GoogleSpannerClient extends DB {
+  private static Logger logger =
+      Logger.getLogger(GoogleSpannerClient.class);
+
+  private Spanner spanner;
+  private DatabaseClient dbClient;
+
+  @Override
+  public void init() throws DBException {
+    String debug = getProperties().getProperty("googlespanner.debug", null);
+    if (null != debug && "true".equalsIgnoreCase(debug)) {
+      logger.setLevel(Level.DEBUG);
+    }
+
+    final String instanceId = getProperties().getProperty(
+        "googlespanner.instance", null);
+    if (instanceId == null) {
+      throw new DBException(
+          "Required property \"googlespanner.instance\" missing.");
+    }
+    final String database = getProperties().getProperty(
+        "googlespanner.database", null);
+    if (database == null) {
+      throw new DBException(
+          "Required property \"googlespanner.database\" missing.");
+    }
+
+    final SpannerOptions options = SpannerOptions.newBuilder().build();
+    this.spanner = options.getService();
+    final DatabaseId db = DatabaseId.of(options.getProjectId(), instanceId, database);
+    this.dbClient = spanner.getDatabaseClient(db);
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    super.cleanup();
+
+    try {
+      this.spanner.closeAsync().get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new DBException(e);
+    }
+  }
+
+  @Override
+  public Status read(String table, String key, Set<String> fields,
+          HashMap<String, ByteIterator> result) {
+    // XXX: Key is not escaped. Unclear to me at this time how to do this in Spanner's SQL dialect.
```

Comment:
Standard escaping problem. Depending on whether or not we want to rely on the workload generating keys that can be naively put between single quotes, this code is wrong and arbitrary SQL may be executed. Whether or not this is acceptable given the intentions of YCSB workload generation I'm not sure, but I timed out on attempting to find out what escaping is required in the Spanner dialect.